src/loader/tsbgw_message_queue.c

```diff
@@ -0,0 +1,273 @@
+static QueueResponseType queue_add(TsbgwMessageQueue * queue, TsbgwMessage * message)
+{
+	QueueResponseType message_sent = QUEUE_FULL;
+	pid_t		reader_pid = InvalidPid;
+
+	LWLockAcquire(queue->lock, LW_EXCLUSIVE);
+	if (queue->num_elements < TSBGW_MAX_MESSAGES)
+	{
+		memcpy(&queue->buffer[(queue->read_upto + queue->num_elements) % TSBGW_MAX_MESSAGES], message, sizeof(TsbgwMessage));
+		queue->num_elements++;
+		message_sent = MESSAGE_SENT;
+		reader_pid = queue->reader_pid;
+	}
+	LWLockRelease(queue->lock);
+
+	if (reader_pid != InvalidPid)
+		SetLatch(&BackendPidGetProc(reader_pid)->procLatch);
+	else
+		message_sent = READER_DETACHED;
+	return message_sent;
+}
+
+static void
+queue_set_reader(TsbgwMessageQueue * queue)
+{
+
+	LWLockAcquire(queue->lock, LW_EXCLUSIVE);
+	if (queue->reader_pid == InvalidPid)
+		queue->reader_pid = MyProcPid;
+	else if (queue->reader_pid != MyProcPid)
+		ereport(ERROR, (errmsg("only one reader for allowed for TimescaleBGW message queue")));
+	LWLockRelease(queue->lock);
+
+}
+
+
+static TsbgwMessage * queue_remove(TsbgwMessageQueue * queue)
+{
+	TsbgwMessage *message = NULL;
+
+	LWLockAcquire(queue->lock, LW_EXCLUSIVE);
+	if (queue->reader_pid != MyProcPid)
+		ereport(ERROR, (errmsg("cannot read if not reader for TimescaleBGW message queue")));
+
+	if (queue->num_elements > 0)
+	{
+		message = palloc(sizeof(TsbgwMessage));
+		memcpy(message, &queue->buffer[queue->read_upto], sizeof(TsbgwMessage));
+		queue->read_upto = (queue->read_upto + 1) % TSBGW_MAX_MESSAGES;
+		queue->num_elements--;
+	}
+	LWLockRelease(queue->lock);
+	return message;
+}
+
+/*construct a message*/
+static TsbgwMessage * tsbgw_message_create(TsbgwMessageType message_type, Oid db_oid)
+{
+	TsbgwMessage *message = palloc(sizeof(TsbgwMessage));
+	dsm_segment *seg;
+
+	seg = dsm_create(TSBGW_ACK_QUEUE_SIZE, 0);
+
+	*message = (TsbgwMessage)
+	{
+		.message_type = message_type,
+			.sender_pid = MyProcPid,
+			.db_oid = db_oid,
+			.ack_dsm_handle = dsm_segment_handle(seg)
+	};
+
+	return message;
+}
+
+/*
+ * write element to queue, wait/error if queue is full
+ * consumes message and deallocates
+ */
+extern bool
+tsbgw_message_send_and_wait(TsbgwMessageType message_type, Oid db_oid)
+{
+	QueueResponseType send_result;
+	shm_mq	   *ack_queue;
+	dsm_segment *seg;
+	shm_mq_handle *ack_queue_handle;
+	Size		bytes_received = 0;
+	bool	   *data = NULL;
+	TsbgwMessage *message;
+	bool		message_sent = false;
+
+	message = tsbgw_message_create(message_type, db_oid);
+
+	seg = dsm_find_mapping(message->ack_dsm_handle);
+	ack_queue = shm_mq_create(dsm_segment_address(seg), TSBGW_ACK_QUEUE_SIZE);
+	shm_mq_set_receiver(ack_queue, MyProc);
+	ack_queue_handle = shm_mq_attach(ack_queue, seg, NULL);
+
+
+	send_result = queue_add(tsbgw_mq, message);
+	if (send_result == MESSAGE_SENT)
+	{
+		shm_mq_wait_for_attach(ack_queue_handle);
+		shm_mq_receive(ack_queue_handle, &bytes_received, (void **) &data, false);
```

Comment:
Seems dangerous to block here on ack. Can't we wait with a timeout on the latch first, like suggested above, then receive with `nowait=true`? 

Should also check return value here in case of, e.g., `SHM_MQ_DETACHED`.